docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+
+### Implementation Prerequisites
+
+1. clusterctl should be able to install CNI. This was discussed on Cluster-API Breakout meeting held on 10/10. [Relevant issue #534](https://github.com/kubernetes-sigs/cluster-api/issues/534)
+
+### Implementation details
+
+The implementation will be based on [Go Templates](https://golang.org/pkg/text/template/). That allows implementers to source environment variables, pass data such as IP addresses, host names, and kubeadm tokens, configure Kubelet and Cloud Provider, without any change to the Cluster-API source code.
+
+Similar approach is used by many projects and providers, including:
+
+* [Cluster-API provider for DigitalOcean](https://github.com/kubermatic/cluster-api-provider-digitalocean/blob/master/cloud/digitalocean/actuators/machine/userdata.go)
+* [Cluster-API provider for GCP](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/master/pkg/cloud/google/metadata.go)
+* [Cluster-API provider for AWS](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/4553a80b6337b4adcc378c07db943772d30fbc78/pkg/cloud/aws/services/ec2/bastion.go)
+* [Cluster-API provider for VSphere](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/cloud/vsphere/provisioner/common/templates.go)
+
+The implementation will be realized using interfaces, meaning that for each supported operating system interface will be implemented. The interface has two functions, for getting user data for Master instances and for workers.
+
+Both functions should take the following arguments:
+
+* Environment variables, provided as a string map and then parsed and appended on the top of script.
+* Kubernetes version, including Kubelet, kubeadm, kubernetes-cni versions.
+* Kubelet and Cloud Provider configuration files
+* To be discussed how we want to provide those:
+* Kubeadm configuration file. Do we want to be able to pass it clusterctl, as a Secret/ConfigMap, or as a function parameter?
+
+```go
+type Userdata interface {
+    MasterUserData(envVars map[string]string, kubeVersion string, cniVersion string) (string, error)
+    NodeUserData(envVars map[string]string, kubeVersion, cniVersion string) (string, error)
+}
+```
+
+### Decisions to be made
+
+* What Kubernetes versions should be supported?
+    * Kubernetes supports 3 version at the same time.
+    * It should be possible to provide scripts for all 3 versions. Usually there are no changes in the bootstrap scripts between older and newer versions.
+* What operating systems should be supported? The following three are supported by most of cloud providers:
+    * Ubuntu 16.04 and Ubuntu 18.04. Very popular and very widely used
+    * Container Linux
+    * CentOS 7
+* What container engine should be set up by the scripts?
+    * Docker - very popular and very widely used in production.
+    * containerd - newer solution, seems to work well. Used by AWS provider.
+    * Opinion: we should choose one and stay with it.
```

Comment:
@sflxn: Where would we keep scripts used to set up other container runtimes? I assume you're not suggesting we just choose one and only support that, right? Strongly isolated container runtimes are coming.

@roberthbailey: Also, even if we pick one today we will need a way to upgrade to a newer version or a different one in the future. For the GCP provider we will want to track what the node folks qualify on GCE/GKE even if that isn't what is being tested on other environments.

@alvaroaleman: The idea here would be to provide something that "just works" and can be used as a default, if ppl decide they want another runtime they can decide to not use that script or add another script template for that other runtime